ddsrouter_core/src/cpp/writer/implementations/rtps/CommonWriter.cpp

```diff
@@ -0,0 +1,356 @@
+        const fastrtps::TopicAttributes& topic_attributes,
+        const fastrtps::WriterQos& writer_qos,
+        const utils::PoolConfiguration& pool_configuration)
+    : BaseWriter(participant_id, topic, payload_pool)
+    , repeater_(repeater)
+    , rtps_participant_(rtps_participant)
+{
+    internal_entities_creation_(
+        history_attributes,
+        writer_attributes,
+        topic_attributes,
+        writer_qos,
+        pool_configuration);
+}
+
+CommonWriter::~CommonWriter()
+{
+    // This variables should be set, otherwise the creation should have fail
+    // Anyway, the if case is used for safety reasons
+
+    // Delete writer
+    if (rtps_writer_)
+    {
+        // Delete the CommonWriter the History is cleaned
+        fastrtps::rtps::RTPSDomain::removeRTPSWriter(rtps_writer_);
+    }
+
+    // Delete History
+    if (rtps_history_)
+    {
+        delete rtps_history_;
+    }
+
+    logInfo(DDSROUTER_RTPS_WRITER, "Deleting CommonWriter created in Participant " <<
+            participant_id_ << " for topic " << topic_);
+}
+
+// Specific enable/disable do not need to be implemented
+utils::ReturnCode CommonWriter::write_(
+        std::unique_ptr<DataReceived>& data) noexcept
+{
+
+    // Take new Change from history
+    fastrtps::rtps::CacheChange_t* new_change = rtps_writer_->new_change(eprosima::fastrtps::rtps::ChangeKind_t::ALIVE);
+
+    // If still is not able to get a change, return an error code
+    if (!new_change)
+    {
+        return utils::ReturnCode::RETCODE_ERROR;
+    }
+
+    // Get the Payload without copy only if it has length
+    if (data->payload.length > 0)
+    {
+        eprosima::fastrtps::rtps::IPayloadPool* payload_owner = payload_pool_.get();
+        if (!payload_pool_->get_payload(data->payload, payload_owner, (*new_change)))
+        {
+            logDevError(DDSROUTER_RTPS_WRITER, "Error getting Payload.");
+            return utils::ReturnCode::RETCODE_ERROR;
+        }
+    }
+
+
+    logDebug(DDSROUTER_RTPS_WRITER,
+            "CommonWriter " << *this << " sending payload " << new_change->serializedPayload << " from " <<
+            data->qos.source_guid);
+
+    // Fill cache change with specific data to send
+    fill_to_send_data_(new_change, data);
+
+    // Get params to write (if set)
+    eprosima::fastrtps::rtps::WriteParams write_params;
+
+    // RPC support
+    // If writer params has been set specifically, use them in change
+    if (data->qos.write_params.is_set())
+    {
+        write_params.related_sample_identity(data->qos.write_params.value.related_sample_identity());
+    }
+
+    // Send data by adding it to CommonWriter History
+    rtps_history_->add_change(new_change, write_params);
+
+    // At this point, write params is now the output of adding change
+    fill_sent_data_(write_params, data);
+
+    // TODO: check if this being synchronous means that we can remove the change in case of volatile and/or best_effort
```

Comment:
Yes, I believe is is safe to remove them.

Reply:
Not sure anymore, we need to first assess the impact on performance for this change.